### ISSUE TITLE
Clippy auto-fix fail on `Option<&mut T>` argument.

### ISSUE THREAD

**Issue by zirconium-n** [rust-lang/rust#89758]
I tried run `cargo clippy --fix` on this code:

```rust

fn main() {
    let mut i = 1;
    let mut bar = Some(&mut i);

    foo(bar.as_deref_mut());
    foo(bar.as_deref_mut());

    println!("Hello, world!");
}


fn foo(_: Option<&mut i32>) {
    todo!()
}
```

This happened: 

```
after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see 
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
warning: variable does not need to be mutable
 --> src/main.rs:4:9
  |
4 |     let mut bar = Some(&mut i);
  |         ----^^^
  |         |
  |         help: remove this `mut`
  |
  = note: `#[warn(unused_mut)]` on by default

error[E0382]: use of moved value: `bar`
 --> src/main.rs:7:9
  |
4 |     let mut bar = Some(&mut i);
  |         ------- move occurs because `bar` has type `std::option::Option<&mut i32>`, which does not implement the `Copy` trait
5 | 
6 |     foo(bar);
  |         --- value moved here
7 |     foo(bar);
  |         ^^^ value used here after move

error: aborting due to previous error; 1 warning emitted

For more information about this error, try `rustc --explain E0382`.
Original diagnostics will follow.

warning: derefed type is same as origin
 --> src/main.rs:6:9
  |
6 |     foo(bar.as_deref_mut());
  |         ^^^^^^^^^^^^^^^^^^ help: try this: `bar`
  |
  = note: `#[warn(clippy::needless_option_as_deref)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_option_as_deref

warning: derefed type is same as origin
 --> src/main.rs:7:9
  |
7 |     foo(bar.as_deref_mut());
  |         ^^^^^^^^^^^^^^^^^^ help: try this: `bar`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_option_as_deref
```

**Comment by JohnTitor** [comment]
The failure is due to Clippy's false positive, could you report it on the Clippy repo (https://github.com/rust-lang/rust-clippy) instead? Closing as it's not a rustc bug.
